pkg/search/postgres/store_cache.go

```diff
@@ -0,0 +1,386 @@
+	checker walker.PermissionChecker,
+) Store[T, PT] {
+	underlyingStore := NewGenericStoreWithPermissionChecker[T, PT](
+		db,
+		schema,
+		pkGetter,
+		insertInto,
+		copyFromObj,
+		setAcquireDBConnDuration,
+		setPostgresOperationDurationTime,
+		checker,
+	)
+	store := &cachedStore[T, PT]{
+		schema:            schema,
+		pkGetter:          pkGetter,
+		permissionChecker: checker,
+		cache:             make(map[string]PT),
+		underlyingStore:   underlyingStore,
+
+		setCacheOperationDurationTime: setCacheOperationDurationTime,
+	}
+	// Initial population of the cache. Make sure it is in sync with the DB.
+	store.repopulateCache()
+	return store
+}
+
+// cachedStore implements subset of Store interface for resources with single ID.
+type cachedStore[T any, PT clonedUnmarshaler[T]] struct {
+	schema                        *walker.Schema
+	pkGetter                      primaryKeyGetter[T, PT]
+	setCacheOperationDurationTime durationTimeSetter
+	permissionChecker             walker.PermissionChecker
+	targetResource                permissions.ResourceMetadata
+	underlyingStore               Store[T, PT]
+	cache                         map[string]PT
+	cacheLock                     sync.RWMutex
+}
+
+// Upsert saves the current state of an object in storage.
+func (c *cachedStore[T, PT]) Upsert(ctx context.Context, obj PT) error {
+	dbErr := c.underlyingStore.Upsert(ctx, obj)
+	if dbErr != nil {
+		return dbErr
+	}
+	defer c.setCacheOperationDurationTime(time.Now(), ops.Upsert)
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	c.addToCacheNoLock(obj.Clone())
+	return nil
+}
+
+// UpsertMany saves the state of multiple objects in the storage.
+func (c *cachedStore[T, PT]) UpsertMany(ctx context.Context, objs []PT) error {
+	dbErr := c.underlyingStore.UpsertMany(ctx, objs)
+	if dbErr != nil {
+		return dbErr
+	}
+	defer c.setCacheOperationDurationTime(time.Now(), ops.UpdateMany)
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	for _, obj := range objs {
+		c.addToCacheNoLock(obj.Clone())
+	}
+	return nil
+}
+
+// Delete removes the object associated to the specified ID from the store.
+func (c *cachedStore[T, PT]) Delete(ctx context.Context, id string) error {
+	obj, found, err := c.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+	if !found {
+		return nil
+	}
+	if !c.isWriteAllowed(ctx, obj) {
+		return nil
+	}
+	dbErr := c.underlyingStore.Delete(ctx, id)
+	if dbErr != nil {
+		return dbErr
+	}
+	defer c.setCacheOperationDurationTime(time.Now(), ops.Remove)
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	delete(c.cache, id)
+	return nil
+}
+
+// DeleteMany removes the objects associated to the specified IDs from the store.
+func (c *cachedStore[T, PT]) DeleteMany(ctx context.Context, identifiers []string) error {
+	objects, _, err := c.GetMany(ctx, identifiers)
+	if err != nil {
+		return err
+	}
+	filteredIDs := make([]string, 0, len(objects))
+	for _, obj := range objects {
+		if !c.isWriteAllowed(ctx, obj) {
```

Comment:
If write isn't allowed on any of the objects, it seems like we should just reject the whole request?

Reply:
@rhybrillou Still need to address this one

Reply:
The behaviour would not be consistent with the current state of the postgres generated store.
There are generated SAC tests. They break when the whole request is rejected.
see https://github.com/stackrox/stackrox/blob/e86687f6dff0ea9d0d95e71bce68fe6696dddc69/tools/generate-helpers/pg-table-bindings/store_test.go.tpl#L422-L448

Reply:
Outside of this PR then, but can you create a JIRA for this? It seems like someone trying to write data they shouldn't should be rejected explicitly and not opaquely

Reply:
https://issues.redhat.com/browse/ROX-21609